### DISCUSSION TITLE
Load multiple semantic model files and use EObjectQuery(eobj).getInverseReferences("field") to find the reference problem.

### DISCUSSION THREAD

**Discussion by pelva** [eclipse-sirius/sirius-desktop#340] [Q&A]
Hi,
I used Eclipse 2022-06 (4.24.0) and Sirius 7.0.1 to create a Diagram to represent the relationships between my semantic models.
Then I opened Diagram programmatically according to the method introduced in the Sirius forum.

> 	/*
	 * - First you will need to create a session programmatically using
	 * SessionFactory.INSTANCE.createSession and then open it (using the open
	 * method). - Next you can add your semantic resources to your new session using
	 * org.eclipse.sirius.business.api.session.Session.addSemanticResource(URI,
	 * IProgressMonitor). - Now, that you have the needed resources in your session,
	 * you will want to activate the Viewpoint you designed on your session. You can
	 * wrap your session in UserSession and use the selectViewpoints method or
	 * directly execute the command ChangeViewpointSelectionCommand. - Finally, you
	 * can create a representation using the API
	 * DialectManager.INSTANCE.createRepresentation.
	 */

How I implement it:
I added a button on the toolbar. 
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/e2c4bb40-56e5-44e1-9681-f83e57e47fda)

After clicking this button, the Session is obtained through SessionManager.INSTANCE.getExistingSession.
If the return value is null, the Session is created through SessionFactory.INSTANCE.createSession, and then the open method is executed. Then traverse the semantic model files in the project and call addSemanticResource in session.getTransactionalEditingDomain().getCommandStack().execute(new RecordingCommand(session.getTransactionalEditingDomain())

Then I search for the Domain Class in the Diagram. **If there are multiple semantic model files, there will be multiple Domain Classes, and I take the first one I find.**
Then use userSession selectViewpoint, DialectManager.INSTANCE.getAvailableRepresentationDescriptions/DialectManager.INSTANCE.createRepresentation
DialectUIManager.INSTANCE.openEditor

Everything works fine if all elements in the semantic model are in one file.
![right](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/58d7c514-addf-4b17-b2bb-385c2e500c66)

The problem is that in actual use, different elements may be located in different semantic model files. For example, one element refers to other elements through proxies.

In my diagram, C1 is the container, a1..a5 is represented as a node in C1, 
a1..a5 is associated with C1 through the middle element, (These elements may all be in different files)
so I need to use EObjectQuery(C1).getInverseReferences("field") to Find the middle element.
When I open Diagram for the first time, getInverseReferences returns an empty list.

![empty](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/79f16716-f9b4-408b-85b7-a0fb3f566e7f)

At this time, I save the editor where the diagram is located and then click the toolbar button again to open the diagram. 
then getInverseReferences returns the correct list and everything is normal.

I checked the source code of EObjectQuery and I think it looks for references through all semantic models.
I'm not sure what I suspect is the problem. Can you give me some advice,thank you!

**Comment by pelva** [comment]
In the introduction above, I omitted some information,  these semantic model files are generated programmatically and then added to the Project.
Every time I regenerate the semantic model file I should use SessionFactory.INSTANCE.createSession instead of getExistingSession and this works fine.

**Comment by lfasani** [comment]
Hello
It looks like a bug.
I am not sure it will do a difference but try using `AddSemanticResourceCommand` instead of your command (`session.getTransactionalEditingDomain().getCommandStack().execute(new AddSemanticResourceCommand)`. It is slightly different.

Otherwise, as a workaround, you may close the session and reopen it just after. (beware to not call open on the same Session instance but use SessionManager instead to get a new Session)

Otherwise, if you have to debug a good entry point is here `org.eclipse.sirius.business.internal.session.danalysis.LocalResourceCollectorCrossReferencer`. this is the adapter that will make the inverseCrossReference work

Regards

**Reply by pelva** [reply]
I apologize for the delayed response, but I wanted to extend my sincere gratitude for taking the time to reply to my query. Your insights were invaluable and greatly appreciated.

Thank you once again for your assistance.

Best regards
